### DISCUSSION TITLE
eeg-notebooks needs a new name. pronto! share your thoughts...

### DISCUSSION THREAD

**Discussion by JohnGriffiths** [NeuroTechX/EEG-ExPy#249] [General]
Hello friends and family of eeg-notebooks.

Gather round. We need to have an important discussion. 

We need a new name for the project, and we'd love your input!

Add your thoughts and ideas in the thread below.
We will need a final decision on this pretty soon - by the end of November.(for reasons explained below).
Also, seems in elaborating on this I accidentally wrote a little history essay, lol. Enjoy! Or, if you find it dull, just skip to the important bit at the bottom. 

---

***Why does it need to change?***

Ultimately, the issue is that the moniker “eeg-notebooks” has been out of date for quite a while. 

This name was originally coined when @jdpigeon Dano (Morrison) embarked on a major refactor of @alexandrebarachant 's  [muse-lsl library](https://github.com/alexandrebarachant/muse-lsl) about 6 years ago, which many of us were using at the time for various side projects. During this work it became clear that the stimulus presentation and data analysis functionality that was originally part of `muse-lsl` was conceptually very separate to (and in fact much more general than) the lsl-based muse eeg data streaming functionality, although the latter much more widely used. So we decided to create a new library, and `eeg-notebooks` was born. 

At that time, eeg-notebooks was *very* heavily centred on jupyter notebooks, which were still a shiny new idea in scientific computing that were taking the world by storm, and not the all-pervasive part of the software furniture they are today. We'd recently shifted from Alexandre's [original approach](https://github.com/alexandrebarachant/muse-lsl/blob/b336f55f1767fc6822a3d16e16654820b44b0ab2/notebooks/P300%20with%20Muse.ipynb) of running recordings separately and loading the results in to notebooks for analysis, to [a fully notebook-based workflow](https://github.com/NeuroTechX/eeg-notebooks_v0.1/blob/v0.1/notebooks/N170.ipynb), where the code for bluetooth eeg device searching, connecting, running stimulus presentation were also included. The primary source code at that time was also the jupyter `.ipynb` files themselves. 

So the library back after the `muse-lsl` split was literally a set of notebooks that did end-to-end pretty much everything involved in running an analyzing an eeg experiment. Which we were pretty buzzed about. And `eeg-notebooks` was indeed a very natural name for this wonderful new entity. 

However, since then the code base and project more broadly has gone through several further evolutions. and is just not as notebook-centric as it originally was. A major one is that there are now no longer any jupyter notebooks in the source code. This is absolutely necessary for a healthy Python software project. Jupyter notebooks, as wonderful as they are, a) do not diff well (making them impossible for multi-contributor edits), and leave large binary blobs in the git history. So, following the lead of many excellent Python libraries such as `mne` and `nliearn` we switched to a sphinx-gallery code base structure, and website builds running on github CI actions, which gives both a nice website and constructs downloadable `.ipynb` files, based on a much more manageable simple text file source code. Another change is that we have come full circle, with streaming, stimulus presentation, and data analysis being [separated in the code](https://neurotechx.github.io/eeg-notebooks/experiments/vn170.html) and workflow documentation. 

As cool as it was, running bluetooth streaming and psychopy from within a notebook caused a lot of headaches and could be quite clunky. Moreover, it is rare that one wants to analyze data collected from just a single experiment run - ideally one run several recording blocks of eg ~5mins each, and load these in together. Which would involve many redundant, repeated lines and stdouts in the data collection part of the code. 

Thus, while `eeg-notebooks` is still in its heart very much still the same library we were using back in 2016, notebooks per se really have very little to do with it. 

And in fact, really, if we're honest, it was always a bit of a distraction. The cool thing about `eeg-notebooks` is the fact that it lets you run cognitive neuroscience experiments on your laptop with cheap mobile eeg devices. That is what we should be focusing on, not the currently used software implementations - which are ofc liable to change again in the future. 

---

***Why now and what's the rush?***

The eventual need for a re-brand has been noted amongst the developer team for a while, but has never been a priority item.
The reason that I'm now bumping it up is that I want in 2024 to be a bit more active getting the word out to various communities about this project - including in particular publishing a software/methodology paper, as well as some scientific conference presentations and posters.
This will be very important for visibility and credibility, as well as the slightly longer-term goal of ideally securing some direct funding for developer time, educational workshops, etc. 
I would like to kick these efforts off with posters at CNS 2024 (Toronto) OHBM 2024 (Seoul) - both of which have submission deadlines of early December. 
The contents of these will be finalized later, but the titles will need to be decided in time for these deadlines. 
So, we need our new name for the project by the end of November. 

---

***What are the candidates?***

We need two things: a library name, and an abbreviated name. They both need to be catchy and reflect at least some of the project's core elements. 

*Candidate 1:*

"CNX: A library for Cognitive Neuroscience Experiments"  
Variant options:  
CNX  
CNE  
CNEx  
CoNeX  
CoNeEx  
CoNeExPy  
CNXPy 
etc.

Pros:    
- This is probably my favoured option as it has the most direct mapping of the main library raison d'etre to acronym letters  

Cons:
- It's quite dull- Doesn't refer to EEG or mobile 
- ...but conversely, we may not want to restrict the scope to EEG and mobile  

*Candidate 2:* 

"NTExpt: A library for NeuroTechnology Experiments 

Variant options:  

NTExpt  
NTE  
NTEx  
NTXpt  
NTXPy
etc.    

Pros:  
- Leans into the neurotechx origin story + community and brand- 'neurotechnology' is inclusive of eeg but not restricted to it  
- Sounds cool  

Cons:  
- Possibly a bit too close to neurotechx ?- Doesn't refer directly to EEG or mobile  
- the 'expt' suffix is a bit verbally clunky

**Comment by danielemarinazzo** [comment]
Exciting times!
I would tend to support the first candidate. I always found the neurotech part a bit weird, even if it spoke to a certain cyberpunk attitude of doing experiments with the low cost devices. And "X" is now having a bad rap somehow.
This said, I do feel the dullness and lack of specificity. Is it good to keep a more general scope in mind, but at the moment is basically all EEG-based, right? And also not clearly a candidate to replace current settings in traditional EEG labs, right? Maybe insist on the "accessible" part? Something like

Python-powered (EEG exploration for) accessible cognitive research?

The part in parenthesis can be dropped if we want to go beyond EEG any time soon.

**Reply by JohnGriffiths** [reply]
Great thoughts @danielemarinazzo, thanks! 

> I would tend to support the first candidate. I always found the neurotech part a bit weird, even if it spoke to a certain cyberpunk attitude of doing experiments with the low cost devices.

Ya we are totes cyberpunk😎

> And "X" is now having a bad rap somehow

Indeed, I feel this. Hopefully Musk hasn't permanently tarnished the NeuroTechX brand with the Twitter rename.

> This said, I do feel the dullness and lack of specificity. Is it good to keep a more general scope in mind, but at the moment is basically all EEG-based, right?

Yes basically it's all eeg. I think it's likely that there will be some muse-like fnirs devices in the picture pretty soon though.   


> And also not clearly a candidate to replace current settings in traditional EEG labs, right?

Yes and no to this. Actually there is no reason why eegnb expts could not be used quite extensively in research eeg labs, as psychopy is used pretty widely. The difference is that you would just run the expts and send (over-wire) triggers, but not control the eeg recording part, as that would be done with some dedicated desktop software. We have some stuff w eegnb expts on the 64chan biosemi started in the lab now. 

Also, AFAIK, there are not any major experiment collection libraries out there for psychopy; ppl seem to mainly hand build from scratch (?) or with the builder, or use misc legacy code.


> Maybe insist on the "accessible" part? Something like
> 
> Python-powered (EEG exploration for) accessible cognitive research?
> 

Also a good option.  

So that would be PACR or ACNE or sth I suppose.



> The part in parenthesis can be dropped if we want to go beyond EEG any time soon.

Truse:

**Comment by JohnGriffiths** [comment]
Let's hear your thoughts ppl - 

@jdpigeon @hubertjb @JadinTredup @ErikBjare @royyannick @tmorshed @retiutut @sbastiaens @p-oveisi

**Comment by ErikBjare** [comment]
I must admit, the eeg-notebooks name has grown on me. I don't like the idea of the new name being an abbreviation.

The project has barely had any commits in the last year, and unless that's changing, I only see minimal value of a name change (arguably not enough to justify it). 

I can't speak for everyone, but I find it really difficult to contribute when one of the dependencies (psychopy) always cause problems in CI and elsewhere. I understand it is needed for the experiments, but it is not for data collection/devices (which I often did without psychopy) nor analysis.

I wish eeg-notebooks could be installed without the monstrous psychopy. This was the driving force between me copying the device files over into my own project for my thesis, instead of using the eeg-notebooks implementation directly. 

It's true that the scope has gone way beyond just "eeg notebooks", but to me it seems the increase in scope is the problem, not the name.

If some parts of the project could be split out into a new project/repo (with a new name to fit the scope), that'd make more sense to me!

But these are just my barely formed thoughts, don't take them too seriously :)

---

Edit: I suppose it means I am in favor of doing what led to the current name to begin with :)

> During this work it became clear that the stimulus presentation and data analysis functionality that was originally part of muse-lsl was conceptually very separate to (and in fact much more general than) the lsl-based muse eeg data streaming functionality, although the latter much more widely used. So we decided to create a new library, and eeg-notebooks was born.

**Reply by JohnGriffiths** [reply]
Ciao Erik. 

Thanks for the thoughtful rant ;) 

a) Good to hear from you, hope you're well!

b) Slow dev for the part year has been in part due to me being on pat leave for the past 9 months. Which has now concluded. Hence this discussion! Ore and Taha especially have been holding the ship very well in my absence nonetheless. But yes there will be more commits hopefully to the level that would justify the name change, as you put it. 

c) Yes of course we all agree on the psychopy headache. As we have discussed before, I think the solution to this should be alternative installation modes, not splitting up the library. I agree this would be a useful update to push through asap.

d) Happy to hear other suggestions of better name choices if the suggested candidates don't float your boat

e) I suppose we could consider the option of keeping the original name and having the extended spiel in my post as the excuse for why we have it as a retro nod to the heady days of 2017. MNE-Python, after all, is named after the source-localization algorithm that is definitely not its main feature any more. But that seems a bit contrived to me personally tbh.

**Reply by ErikBjare** [reply]
I think you are right on all points. 

After looking over everything, including the develop branch (that I missed/forgot about, where CI is passing apart from typechecking), and refreshing my memory: I take back much of what I said.

Yes, we should be able to set up a extras group that includes psychopy which makes the whole thing a lot easier to depend on.

I don't have any great ideas for names, but here is my scratchpad:

 - eeg-experiments
 - EEGEx / EEGExPy (lowercase `eegexpy`)
 - NeuroTechEx (as a funny nod to NeuroTechX, but too similar to be practical)
 - EEGxperiments

I think trying to force in "cognitive neuroscience" might make it less clear what it's about at a glance, even if perhaps technically most accurate.

I'm all for changing the name, but given I think the current name isn't all that bad, my bar for a better name might be a bit higher :)

**Comment by royyannick** [comment]
What a journey it's been!

I don't have a strong opinion on this. I somewhat agree that the name ain't
that bad and has grown on me too. However, the project certainly has
evolved from the initial Quick and Dirty Notebooks vision.

I would suggest keeping the EEG part in the name as I don't think we'll be
pivoting away from that for a while and it's useful/powerful/SEO friendly.
I would agree to not try to go "too wide" with the name, unless the
ambitions and a realistic roadmap take the project there.

I think the first question challenge I'd have is "where do we see the
project evolving?" for example, JG you said:
------------------------------------------------
*[...]*
*Cons:*

   - *It's quite dull- Doesn't refer to EEG or mobile *
   - *...but conversely, we may not want to restrict the scope to EEG and
   mobile*

*Candidate 2: *

*"NTExpt: A library for NeuroTechnology Experiments*
------------------------------------------------

As much as I love anything NT/NT(X) like, I'm biased obviously NTX <3, I'm
not sure this would be a good representation unless we see the project
expanding strongly beyond EEG.

I'd be more aligned with something closer to your first suggestions while
trying to keep the EEG if it fits well and we're not pushing it too hard.
(e.g. CoNeeGX)

-Yannick.

**Comment by JohnGriffiths** [comment]
Thanks for the excellent feedback so far folks.

From the input and some further discussions with others I have a new preferred option, that slightly different to those listed so far:

"EEG-ExPy: EEG Experiments in Python"

```python
from eegexpy.experiments import p300
```

_Notes:_

- Very clear what its about

- Has a small homage to `eeg-notebooks` (which has clearly grown on ppl a little) with the `eeg-` prefix

- Includes `eeg` , which I agree is preferable from a "branding" pov

- Also agree that trying to shoehorn "Cognitive Neuroscience " in might be not be a good idea 

- We could in the future if we care switch to "NTX-ExPy" ("Neurotechnology Experiments in Python"), if useful later break that down into "EEG-ExPy" and "FNIRS-ExPy" in a semi-hierarchical way. MNE is doing this a bit with `mne-nirs`.

- Doing "Ex..." rather than "X..." helps get round the musk brand car crash thing, and also reads naturally as "Experiment"

**Comment by royyannick** [comment]
I love it!

*eegexpy*

It has everything in the name and it's not overly complicated nor too wide.
When Googling it it doesn't clash with anything (only did a 10 sec search,
might want to dig slightly more before committing to be certain, maybe
looking on github as well)
When looking on PyPi it's available.

Thanks John for taking the feedback and moving things forward. Let's see
what the others are saying, but *that one clearly has my vote* :)

-Yannick.

**Comment by JohnGriffiths** [comment]
OK - **EEG-ExPy** it is. 

Last call for objectors! 

Currently drafting [abstract for OHBM 2024](https://docs.google.com/document/d/1CLeuuYbzhhbjaY4Trl9PwNc_0DdHQ9ZdHvTiuFs4SFc/edit?usp=sharing) - due 1st december. 

Feel free to drop in w comments and edits!

**Reply by ErikBjare** [reply]
You won't hear any objections from me, given it was [my suggestion](https://github.com/NeuroTechX/eeg-notebooks/discussions/249#discussioncomment-7633206) 😉

**Reply by retiutut** [reply]
Also, no objections. Been following this thread but didn't have a strong opinion.

**Comment by tmorshed** [comment]
Sketched a new logo on Canva while on the train :)

**To contribute:**
Find the project [here](https://www.canva.com/design/DAF29-Afodk/GeOudPZjxvUlE63F85bPPQ/edit?utm_content=DAF29-Afodk&utm_campaign=designshare&utm_medium=link2&utm_source=sharebutton) and feel free to contribute / make edits.

**Here's a preview:**

![EEG-ExPy](https://github.com/NeuroTechX/EEG-ExPy/assets/14872591/c41d391e-62e7-423e-95c2-02afcbfdfd0b)

**Reply by retiutut** [reply]
Nice!!!
